tools/src/main/scala/org/scalanative/bindgen/Bindgen.scala

```diff
@@ -6,141 +6,58 @@ import scala.collection.immutable.Seq
-                                header: Option[File] = None,
-                                name: Option[String] = None,
-                                packageName: Option[String] = None,
-                                excludePrefix: Option[String] = None,
-                                extraArg: Seq[String] = Seq.empty,
-                                extraArgBefore: Seq[String] = Seq.empty)
-      extends Bindgen {
-
-    def bindgenExecutable(executable: File): Bindgen = {
-      require(executable.exists(), s"Executable does not exist: $executable")
-      copy(executable = Option(executable))
-    }
-
-    def header(header: File): Bindgen = {
-      require(header.exists(), s"Header file does not exist: $header")
-      copy(header = Option(header))
-    }
-
-    def link(library: String): Bindgen = {
-      require(library.nonEmpty, "Library must be non-empty")
-      copy(library = Option(library))
-    }
-
-    def name(name: String): Bindgen = {
-      require(name.nonEmpty, "Name must be non-empty")
-      copy(name = Option(name))
-    }
-
-    def packageName(packageName: String): Bindgen = {
-      require(packageName.nonEmpty, "Package name must be non-empty")
-      copy(packageName = Option(packageName))
-    }
-
-    def excludePrefix(prefix: String): Bindgen = {
-      require(prefix.nonEmpty, "Exclude prefix must be non-empty")
-      copy(excludePrefix = Option(prefix))
+  def generate(options: BindingOptions): Either[Seq[String], Bindings] = {
+    options match {
+      case impl: BindingOptions.Impl =>
+        import impl._
+
+        require(executable.exists, "The executable does not exist")
+
+        val nameOrLibrary = name.orElse(library)
+        require(nameOrLibrary.isDefined,
+                "Name must be specified when no library name is given")
+
+        def withArgs(arg: String, values: Iterable[String]) =
+          values.toSeq.flatMap(Seq(arg, _))
+
+        val options =
+          withArgs("--name", nameOrLibrary) ++
+            withArgs("--link", library) ++
+            library.fold(Seq("--no-link"))(_ => Seq.empty) ++
+            withArgs("--package", packageName) ++
+            withArgs("--exclude-prefix", excludePrefix) ++
+            withArgs("--extra-arg", extraArgs) ++
+            withArgs("--extra-arg-before", extraArgsBefore) ++
+            Seq(header.getAbsolutePath, "--")
+
+        val cmd    = Seq(executable.getAbsolutePath) ++ options
+        val stdout = ListBuffer[String]()
+        val stderr = ListBuffer[String]()
+        val nl     = System.lineSeparator()
+        val logger = ProcessLogger(stdout.+=, stderr.+=)
+
+        Process(cmd).!(logger) match {
+          case 0 =>
+            Right(
+              new Bindings(nameOrLibrary.get,
+                           stdout.mkString(nl),
+                           Seq(stderr: _*)))
+          case _ => Left(Seq(stderr: _*))
+        }
+
+      case _ => Left(Seq("Illtyped BindingOptions"))
```

Comment:
The same question: is it possible?

Reply:
It should not be possible, but I'm a defensive programmer and this seems to be the safer option than just letting Scala throw an exception of its choosing.